### PR TITLE
Debug info demangling workaround

### PR DESCRIPTION
**Pull request by tempdragon**
# TL;DR
This series of commit is to provide a temporary workaround for demangling of cg_gcc.  It will enable the demangling in GDB for backtrace & function names but DISABLE the mangled name printing even with `set print demangle off`.

# Dependency
Required BY rust-lang/gccjit.rs#31

# Details
This part of code will be used even in the final version that supports both demangled and mangled names.

Currently this will disable the generation of mangled name information (just for GCCJIT, if the option `GCC_JIT_BOOL_OPTION_MANGLED_FUNCTION_NAME` is set, making it of NO overhead for those who don't use this option.)
(In my case, when co-generating both the mangled name(as `DW_AT_linkage_name`) and the function name (as `DW_AT_name`) are generated, the GDB will only print the DW_AT_name for a backtrace, despite the DW_AT_lang. However, this problem doesn't turn up in LLVM-generated code).
 
Further research and research is still to be done over how exactly GDB decides whether to demangle the functions names.

### CODE REVIEW

**Comment by antoyo** [comment]
Thanks!

Could you please explain why you decided to go for a function attribute instead of simply adding a new API `gcc_jit_function_set_short_name`?

**Comment by tempdragon** [comment]
Interestingly, I never thought about this. But having heard of your idea, I think I avoided this because the tree node data type does not currently have a field for short names. To keep everything accessible until the di gen stage, The two apparent ways are either to create a new field in `tree` or to keep it in an attribute as I currently do it. It seems that to access it later in the di gen phase, I cannot avoid both of them easily. Of course there is a third way which is to demangle the name in gccjit, consuming less space and potentially more time(?).

**Comment by antoyo** [comment]
I would be very interested in seeing how the other frontends do this.
I'll take a look.

**Comment by tempdragon** [comment]
LLVM generates it directly.
GCCRS seems to set the short name directly as the names(It looks way too complicated to me to tailor cg_gcc names to something like this, which I didn't succeed in. Maybe it makes sense to use the complete name as a function name and the passed function name as a `DECL_ASSEMBLER_NAME`).

**Comment by tempdragon** [comment]
Another problem in renaming the tree is, the LLVM used a (short name, mangled name)->(DW_AT_name, DW_AT_linkage_name) mapping and the short name they use is effectively shorter than the complete name I am using in the series. I failed to rename the tree (to the shorter names)while getting it compiled without panicking even with the `DECL_ASSEMBLER_NAME` set.

**Comment by tempdragon** [comment]
It may still be possibe, however, to provide an API of `set_short_name()` but still store it as an attribute internally. In this way we can avoid exposure of internal representation while allowing users to set the short name.

**Comment by tempdragon** [comment]
# Updates
> I would be very interested in seeing how the other frontends do this. I'll take a look.

It occurred to me that I should probably set the language name to `"GNU Rust"` and every function with a `DECL_ASSEMBLER_NAME` to `TREE_PUBLIC` to have it correctly demangled in `dwarf2out.cc` (which emits `DW_AT_linkage_name` only when it is PUBLIC("linkable" externally), as is done in `GCCRS`. 

Since publishing all nodes my result in undesirable side effect to program, this is not necessary a good idea. In my own version of code, I added another condition to emit `DW_AT_linkage_name` even if the tree node is not "public", to avoid the potential problem. However, I am not sure which idea is better.

I guess my previous failure of (probably not) doing the same thing to enable demangling in backtrace was that due to:
1. I didn't use "GNU Rust" but "Rust" for the name;
2. I was on a wrong fork of my own code.

# Known Issues
In my editor(Emacs), if the language is set to "GNU Rust", GDB local resolution fails with all locals trailed with many spaces. Some structs are shown with their structures printed. I am not sure whether this has any effects on other editors.

**Comment by tempdragon** [comment]
To verify/evaluate the latest situation, pull this gcc fork and set the `LANG_HOOKS_NAME` to "GNU Rust" from "libgccjit" in `dummy-frontend.cc`.

**Comment by antoyo** [comment]
Just so you know, it make take me a couple of days before I can take a look.

**Comment by tempdragon** [comment]
It doesn't matter. It is always up to you to decide your time to take a look. But if you find my comments annoying, I will reduce my comments.

**Comment by antoyo** [comment]
No, it's OK, you can continue commenting.
I just wanted to let you know that I cannot review right now :) .

**Comment by tempdragon** [comment]
I guess I misunderstood something and I need further investigation now.

**Comment by antoyo** [comment]
Sorry, I completely forgot about this issue (you can ping me in these cases :) ).

What's the status of this? Are you waiting for my review?

**Comment by tempdragon** [comment]
No, I think I should close this and for personal reasons I will have to wait a long time(a yr or so?) before I can really return to cg_gcc work. Thank you so much for your work in cg_gcc.

**Comment by antoyo** [comment]
Ok, I understand. Thank you to your contributions.

Is this something I could reuse if I want to implement this myself?

**Comment by tempdragon** [comment]
Probably not. I can't remenber many details in this branch and probably the same process and ideas should be repeated for someone to implement this feature. The successor to this module will probably need to debug gdb someday in the future as I did... This branch is for reference only.

If anyone is interested in taking over this part of job, FYI, the point is to create a new data structure for debug namspace(Otherwise the gccjit context should be used and we will have to rewrite a lot of code) for correct demangling of each level of module. GDB uses a tree walking strategy rather than demangling to generate the full path and then the debug name of each symbol. (Based on what I read from GDB code.) You need to rebuild the whole module dbg name info hierachy on the cg_gcc then.
